master/all/man/man3/KbBindWidget.3
.TH "KbBindWidget" 3 "Sat Jan 20 2018" "Version v0.2.8 at branch master" "ckb-next" \" -*- nroff -*-
.ad l
.nh
.SH NAME
KbBindWidget \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <src/ckb/kbbindwidget\&.h>\fP
.PP
Inherits \fBQWidget\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBKbBindWidget\fP (\fBQWidget\fP *parent=0)"
.br
.ti -1c
.RI "\fB~KbBindWidget\fP ()"
.br
.ti -1c
.RI "void \fBsetBind\fP (\fBKbBind\fP *newBind, \fBKbProfile\fP *newProfile)"
.br
.in -1c
.SS "Private Slots"

.in +1c
.ti -1c
.RI "void \fBupdateBind\fP ()"
.br
.ti -1c
.RI "void \fBnewLayout\fP ()"
.br
.ti -1c
.RI "void \fBnewSelection\fP (QStringList selection)"
.br
.ti -1c
.RI "void \fBupdateSelDisplay\fP ()"
.br
.ti -1c
.RI "void \fBon_resetButton_clicked\fP ()"
.br
.ti -1c
.RI "void \fBon_copyButton_clicked\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBUi::KbBindWidget\fP * \fBui\fP"
.br
.ti -1c
.RI "\fBKbBind\fP * \fBbind\fP"
.br
.ti -1c
.RI "\fBKbProfile\fP * \fBprofile\fP"
.br
.ti -1c
.RI "QStringList \fBcurrentSelection\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 12 of file kbbindwidget\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "KbBindWidget::KbBindWidget (\fBQWidget\fP *parent = \fC0\fP)\fC [explicit]\fP"

.PP
Definition at line 7 of file kbbindwidget\&.cpp\&.
.PP
References Ui_KbBindWidget::keyWidget, newSelection(), KeyWidget::rgbMode(), Ui_KbBindWidget::setupUi(), and ui\&.
.PP
.nf
7                                           :
8     QWidget(parent),
9     ui(new Ui::KbBindWidget), bind(0), profile(0)
10 {
11     ui->setupUi(this);
12     ui->keyWidget->rgbMode(false);
13     connect(ui->keyWidget, SIGNAL(selectionChanged(QStringList)), this, SLOT(newSelection(QStringList)));
14 }
.fi
.SS "KbBindWidget::~KbBindWidget ()"

.PP
Definition at line 16 of file kbbindwidget\&.cpp\&.
.PP
References ui\&.
.PP
.nf
16                            {
17     delete ui;
18 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void KbBindWidget::newLayout ()\fC [private]\fP, \fC [slot]\fP"

.PP
Definition at line 50 of file kbbindwidget\&.cpp\&.
.PP
References bind, Ui_KbBindWidget::keyWidget, KeyWidget::map(), KbBind::map(), profile, Ui_KbBindWidget::rbWidget, RebindWidget::setBind(), ui, and updateSelDisplay()\&.
.PP
Referenced by setBind()\&.
.PP
.nf
50                             {
51     ui->keyWidget->map(bind->map());
52     ui->rbWidget->setBind(bind, profile);
53     updateSelDisplay();
54 }
.fi
.SS "void KbBindWidget::newSelection (QStringListselection)\fC [private]\fP, \fC [slot]\fP"

.PP
Definition at line 56 of file kbbindwidget\&.cpp\&.
.PP
References currentSelection, Ui_KbBindWidget::rbWidget, RebindWidget::setSelection(), ui, and updateSelDisplay()\&.
.PP
Referenced by KbBindWidget(), and setBind()\&.
.PP
.nf
56                                                     {
57     currentSelection = selection;
58     ui->rbWidget->setSelection(selection, true);
59     updateSelDisplay();
60 }
.fi
.SS "void KbBindWidget::on_copyButton_clicked ()\fC [private]\fP, \fC [slot]\fP"

.PP
Definition at line 103 of file kbbindwidget\&.cpp\&.
.PP
References KbBind::action(), bind, KbMode::bind(), KeyMap::count(), KbProfile::currentMode(), currentSelection, KbBind::isMouse(), KeyMap::keys(), KbBind::map(), KbProfile::modes(), profile, ModeSelectDialog::selection(), KbBind::setAction(), and KeyAction::type()\&.
.PP
.nf
103                                         {
104     QStringList selection = currentSelection;
105     const KeyMap& map = bind->map();
106     int count = selection\&.count();
107     QString type = bind->isMouse() ? "button" : "key";
108     QString text = tr("%1 %2")\&.arg(count)\&.arg(type) + (count == 1 ? "" : "s");
109     if(count == 0){
110         // Copy all keys if none selected
111         selection = map\&.keys();
112         text = tr("all %1s")\&.arg(type);
113     }
114     text = "Copy binding for " + text + " to:";
115     // Display popup
116     ModeSelectDialog dialog(this, profile->currentMode(), profile->modes(), text);
117     if(dialog\&.exec() != QDialog::Accepted)
118         return;
119     // Copy selected keys to selected modes
120     QList<KbMode*> selectedModes = dialog\&.selection();
121     foreach(KbMode* mode, selectedModes){
122         KbBind* modeBind = mode->bind();
123         foreach(const QString& key, selection){
124             modeBind->setAction(key, bind->action(key));
125         }
126     }
127 }
.fi
.SS "void KbBindWidget::on_resetButton_clicked ()\fC [private]\fP, \fC [slot]\fP"

.PP
Definition at line 82 of file kbbindwidget\&.cpp\&.
.PP
References bind, KeyMap::count(), currentSelection, KbBind::isMouse(), KeyMap::keys(), KbBind::map(), KbBind::resetAction(), KeyAction::type(), and updateBind()\&.
.PP
.nf
82                                          {
83     QStringList selection = currentSelection;
84     const KeyMap& map = bind->map();
85     if(selection\&.isEmpty())
86         // Reset all keys if none selected
87         selection = map\&.keys();
88     uint count = selection\&.count();
89     QString type = bind->isMouse() ? "button" : "key";
90     QString text;
91     if(count == map\&.count())
92         text = "<center>Reset all %1s to default?</center>";
93     else if(count == 1)
94         text = "<center>Reset this %1 to default?</center>";
95     else
96         text = tr("<center>Reset %1 %2s to default?</center>")\&.arg(count);
97     if(QMessageBox(QMessageBox::NoIcon, "Confirm action", text\&.arg(type), QMessageBox::Yes | QMessageBox::No, this)\&.exec() != QMessageBox::Yes)
98         return;
99     bind->resetAction(selection);
100     updateBind();
101 }
.fi
.SS "void KbBindWidget::setBind (\fBKbBind\fP *newBind, \fBKbProfile\fP *newProfile)"

.PP
Definition at line 20 of file kbbindwidget\&.cpp\&.
.PP
References bind, KeyWidget::clearSelection(), Ui_KbBindWidget::keyWidget, newLayout(), newSelection(), profile, Ui_KbBindWidget::rbWidget, RebindWidget::setBind(), ui, and updateBind()\&.
.PP
Referenced by KbWidget::modeChanged()\&.
.PP
.nf
20                                                                 {
21     ui->keyWidget->clearSelection();
22     ui->rbWidget->setBind(newBind, newProfile);
23     newSelection(QStringList());
24     if(bind == newBind)
25         return;
26     if(bind){
27         disconnect(bind, SIGNAL(layoutChanged()), this, SLOT(newLayout()));
28         disconnect(bind, SIGNAL(updated()), this, SLOT(updateBind()));
29     }
30     connect(newBind, SIGNAL(layoutChanged()), this, SLOT(newLayout()));
31     connect(newBind, SIGNAL(updated()), this, SLOT(updateBind()));
32     bind = newBind;
33     profile = newProfile;
34     newLayout();
35     ui->rbWidget->setBind(bind, profile);
36     updateBind();
37 }
.fi
.SS "void KbBindWidget::updateBind ()\fC [private]\fP, \fC [slot]\fP"

.PP
Definition at line 39 of file kbbindwidget\&.cpp\&.
.PP
References KbBind::action(), bind, KeyWidget::bindMap(), currentSelection, KeyMap::keys(), Ui_KbBindWidget::keyWidget, KbBind::map(), Ui_KbBindWidget::rbWidget, RebindWidget::setSelection(), ui, and updateSelDisplay()\&.
.PP
Referenced by on_resetButton_clicked(), and setBind()\&.
.PP
.nf
39                              {
40     const KeyMap& map = bind->map();
41     // Build the action list based on the keymap
42     QHash<QString, QString> actions;
43     foreach(const QString& key, map\&.keys())
44         actions[key] = bind->action(key);
45     ui->keyWidget->bindMap(actions);
46     ui->rbWidget->setSelection(currentSelection);
47     updateSelDisplay();
48 }
.fi
.SS "void KbBindWidget::updateSelDisplay ()\fC [private]\fP, \fC [slot]\fP"

.PP
Definition at line 62 of file kbbindwidget\&.cpp\&.
.PP
References bind, currentSelection, KbBind::friendlyActionName(), Key::friendlyName(), KbBind::isMouse(), KbBind::map(), Ui_KbBindWidget::selectLabel, and ui\&.
.PP
Referenced by newLayout(), newSelection(), and updateBind()\&.
.PP
.nf
62                                    {
63     int count = currentSelection\&.count();
64     if(count == 0){
65         // No keys selected
66         ui->selectLabel->setText("Click to select");
67         return;
68     }
69     if(count == 1){
70         // Single key selected: show key name and binding
71         QString key = currentSelection[0];
72         const Key& pos = bind->map()[key];
73         if(!pos)
74             ui->selectLabel->setText("(Unknown)");
75         else
76             ui->selectLabel->setText(pos\&.friendlyName(false)\&.split("\n")[0] + " → " + bind->friendlyActionName(key)\&.split("\n")[0]);
77         return;
78     }
79     ui->selectLabel->setText(QString("%1 %2 selected")\&.arg(count)\&.arg(bind->isMouse() ? "buttons" : "keys"));
80 }
.fi
.SH "Field Documentation"
.PP 
.SS "\fBKbBind\fP* KbBindWidget::bind\fC [private]\fP"

.PP
Definition at line 34 of file kbbindwidget\&.h\&.
.PP
Referenced by newLayout(), on_copyButton_clicked(), on_resetButton_clicked(), setBind(), updateBind(), and updateSelDisplay()\&.
.SS "QStringList KbBindWidget::currentSelection\fC [private]\fP"

.PP
Definition at line 36 of file kbbindwidget\&.h\&.
.PP
Referenced by newSelection(), on_copyButton_clicked(), on_resetButton_clicked(), updateBind(), and updateSelDisplay()\&.
.SS "\fBKbProfile\fP* KbBindWidget::profile\fC [private]\fP"

.PP
Definition at line 35 of file kbbindwidget\&.h\&.
.PP
Referenced by newLayout(), on_copyButton_clicked(), and setBind()\&.
.SS "\fBUi::KbBindWidget\fP* KbBindWidget::ui\fC [private]\fP"

.PP
Definition at line 32 of file kbbindwidget\&.h\&.
.PP
Referenced by KbBindWidget(), newLayout(), newSelection(), setBind(), updateBind(), updateSelDisplay(), and ~KbBindWidget()\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for ckb-next from the source code\&.


src/ckb/kbbindwidget.cpp
#include <QMessageBox>
#include "kbbindwidget.h"
#include "modeselectdialog.h"
#include "rebindwidget.h"
#include "ui_kbbindwidget.h"

KbBindWidget::KbBindWidget(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::KbBindWidget), bind(0), profile(0)
{
    ui->setupUi(this);
    ui->keyWidget->rgbMode(false);
    connect(ui->keyWidget, SIGNAL(selectionChanged(QStringList)), this, SLOT(newSelection(QStringList)));
}

KbBindWidget::~KbBindWidget(){
    delete ui;
}

void KbBindWidget::setBind(KbBind* newBind, KbProfile* newProfile){
    ui->keyWidget->clearSelection();
    ui->rbWidget->setBind(newBind, newProfile);
    newSelection(QStringList());
    if(bind == newBind)
        return;
    if(bind){
        disconnect(bind, SIGNAL(layoutChanged()), this, SLOT(newLayout()));
        disconnect(bind, SIGNAL(updated()), this, SLOT(updateBind()));
    }
    connect(newBind, SIGNAL(layoutChanged()), this, SLOT(newLayout()));
    connect(newBind, SIGNAL(updated()), this, SLOT(updateBind()));
    bind = newBind;
    profile = newProfile;
    newLayout();
    ui->rbWidget->setBind(bind, profile);
    updateBind();
}

void KbBindWidget::updateBind(){
    const KeyMap& map = bind->map();
    // Build the action list based on the keymap
    QHash<QString, QString> actions;
    foreach(const QString& key, map.keys())
        actions[key] = bind->action(key);
    ui->keyWidget->bindMap(actions);
    ui->rbWidget->setSelection(currentSelection);
    updateSelDisplay();
}

void KbBindWidget::newLayout(){
    ui->keyWidget->map(bind->map());
    ui->rbWidget->setBind(bind, profile);
    updateSelDisplay();
}

void KbBindWidget::newSelection(QStringList selection){
    currentSelection = selection;
    ui->rbWidget->setSelection(selection, true);
    updateSelDisplay();
}

void KbBindWidget::updateSelDisplay(){
    int count = currentSelection.count();
    if(count == 0){
        // No keys selected
        ui->selectLabel->setText("Click to select");
        return;
    }
    if(count == 1){
        // Single key selected: show key name and binding
        QString key = currentSelection[0];
        const Key& pos = bind->map()[key];
        if(!pos)
            ui->selectLabel->setText("(Unknown)");
        else
            /// \htmlonly
            ui->selectLabel->setText(pos.friendlyName(false).split("\n")[0] + " → " + bind->friendlyActionName(key).split("\n")[0]); ///> e2 86 92 is keycode for the rightarrow
            /// \endhtmlonly
            /// \latexonly
            /// unprintable code, please refer to html view
            /// \endlatexonly
        return;
    }
    ui->selectLabel->setText(QString("%1 %2 selected").arg(count).arg(bind->isMouse() ? "buttons" : "keys"));
}

void KbBindWidget::on_resetButton_clicked(){
    QStringList selection = currentSelection;
    const KeyMap& map = bind->map();
    if(selection.isEmpty())
        // Reset all keys if none selected
        selection = map.keys();
    uint count = selection.count();
    QString type = bind->isMouse() ? "button" : "key";
    QString text;
    if(count == map.count())
        text = "<center>Reset all %1s to default?</center>";
    else if(count == 1)
        text = "<center>Reset this %1 to default?</center>";
    else
        text = tr("<center>Reset %1 %2s to default?</center>").arg(count);
    if(QMessageBox(QMessageBox::NoIcon, "Confirm action", text.arg(type), QMessageBox::Yes | QMessageBox::No, this).exec() != QMessageBox::Yes)
        return;
    bind->resetAction(selection);
    updateBind();
}

void KbBindWidget::on_copyButton_clicked(){
    QStringList selection = currentSelection;
    const KeyMap& map = bind->map();
    int count = selection.count();
    QString type = bind->isMouse() ? "button" : "key";
    QString text = tr("%1 %2").arg(count).arg(type) + (count == 1 ? "" : "s");
    if(count == 0){
        // Copy all keys if none selected
        selection = map.keys();
        text = tr("all %1s").arg(type);
    }
    text = "Copy binding for " + text + " to:";
    // Display popup
    ModeSelectDialog dialog(this, profile->currentMode(), profile->modes(), text);
    if(dialog.exec() != QDialog::Accepted)
        return;
    // Copy selected keys to selected modes
    QList<KbMode*> selectedModes = dialog.selection();
    foreach(KbMode* mode, selectedModes){
        KbBind* modeBind = mode->bind();
        foreach(const QString& key, selection){
            modeBind->setAction(key, bind->action(key));
        }
    }
}
